File: core/CourseSchedule/CourseSchedule.py
from typing import List


def canFinish(numCourses: int, prerequisites: List[List[int]]) -> bool:
    graph, count = {}, [0] * numCourses
    for end, start in prerequisites:
        count[end] += 1
        if start not in graph:
            graph[start] = []
        graph[start].append(end)

    queue = []
    for i in range(numCourses):
        if count[i] == 0:
            queue.append(i)

    while queue:
        item = queue.pop()
        if item in graph:
            for node in graph[item]:
                count[node] -= 1
                if count[node] == 0:
                    queue.append(node)

    for i in range(numCourses):
        if count[i] != 0:
            return False

    return True

File: core/CourseSchedule/CourseSchedule.java
package com.dbc;

import java.util.*;

public class CourseSchedule {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        int[] count = new int[numCourses];
        Arrays.fill(count, 0);
        for (int[] prerequisite : prerequisites) {
            count[prerequisite[0]]++;
            if (!graph.containsKey(prerequisite[1])) graph.put(prerequisite[1], new ArrayList<>());
            graph.get(prerequisite[1]).add(prerequisite[0]);
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (count[i] == 0) queue.add(i);
        }

        while (!queue.isEmpty()) {
            int item = queue.poll();
            if (graph.containsKey(item)) {
                for (int node : graph.get(item)) {
                    count[node]--;
                    if (count[node] == 0) queue.add(node);
                }
            }
        }

        for (int i = 0; i < numCourses; i++) {
            if (count[i] != 0) return false;
        }
        return true;
    }
}
